contracts/src/token/erc721/extensions/consecutive.rs

```diff
@@ -0,0 +1,1007 @@
+//!
+//! IMPORTANT: Consecutive mint of [`Erc721Consecutive`] tokens is only allowed
+//! inside the contract's Solidity constructor.
+//! As opposed to the Solidity implementation of Consecutive, there is no
+//! restriction on the [`Erc721Consecutive::_update`] function call since it is
+//! not possible to call a Rust function from the Solidity constructor.
+
+use alloc::vec;
+
+use alloy_primitives::{uint, Address, U256};
+use alloy_sol_types::sol;
+use stylus_proc::{external, sol_storage, SolidityError};
+use stylus_sdk::{
+    abi::Bytes, call::MethodError, evm, msg, prelude::TopLevelStorage,
+};
+
+use crate::{
+    token::{
+        erc721,
+        erc721::{
+            Approval, ERC721IncorrectOwner, ERC721InvalidApprover,
+            ERC721InvalidReceiver, ERC721InvalidSender, ERC721NonexistentToken,
+            Erc721, IErc721, Transfer,
+        },
+    },
+    utils::{
+        math::storage::{AddAssignUnchecked, SubAssignUnchecked},
+        structs::{
+            bitmap::BitMap,
+            checkpoints,
+            checkpoints::{Trace160, U96},
+        },
+    },
+};
+
+sol_storage! {
+    /// State of an [`Erc721Consecutive`] token.
+    pub struct Erc721Consecutive {
+        /// Erc721 contract storage.
+        Erc721 erc721;
+        /// Checkpoint library contract for sequential ownership.
+        Trace160 _sequential_ownership;
+        /// BitMap library contract for sequential burn of tokens.
+        BitMap _sequential_burn;
+        /// Used to offset the first token id in
+        /// [`Erc721Consecutive::_next_consecutive_id`].
+        uint96 _first_consecutive_id;
+        /// Maximum size of a batch of consecutive tokens. This is designed to limit
+        /// stress on off-chain indexing services that have to record one entry per
+        /// token, and have protections against "unreasonably large" batches of
+        /// tokens.
+        uint96 _max_batch_size;
+    }
+}
+
+sol! {
+    /// Emitted when the tokens from `from_token_id` to `to_token_id` are transferred from `from_address` to `to_address`.
+    ///
+    /// * `from_token_id` - First token being transferred.
+    /// * `to_token_id` - Last token being transferred.
+    /// * `from_address` - Address from which tokens will be transferred.
+    /// * `to_address` - Address where the tokens will be transferred to.
+    event ConsecutiveTransfer(
+        uint256 indexed from_token_id,
+        uint256 to_token_id,
+        address indexed from_address,
+        address indexed to_address
+    );
+}
+
+sol! {
+    /// Batch mint is restricted to the constructor.
+    /// Any batch mint not emitting the [`Transfer`] event outside of the constructor
+    /// is non ERC-721 compliant.
+    #[derive(Debug)]
+    #[allow(missing_docs)]
+    error ERC721ForbiddenBatchMint();
+
+    /// Exceeds the max number of mints per batch.
+    #[derive(Debug)]
+    #[allow(missing_docs)]
+    error ERC721ExceededMaxBatchMint(uint256 batchSize, uint256 maxBatch);
```

Comment:
We use sneak_case for error's attributes, check https://github.com/OpenZeppelin/rust-contracts-stylus/blob/main/contracts/src/token/erc721/mod.rs#L54-L120